### ISSUE TITLE
Optimization of projection code (comparision with pyresample)

### ISSUE THREAD

**Issue by jmozmoz** [SciTools/cartopy#476]
There was a [discussion on stackoverflow](https://stackoverflow.com/questions/25194017/differences-in-projecting-satellite-images-with-cartopy-and-pyresample) about the differences between cartopy and pyresample. Especially at the moment pyresample is much faster than cartopy.

> > 1. pyresample is much faster than cartopy (depending on the size of the output image up to a factor of 10)
> 
> The cartopy reprojection functionality hasn't been optimized in any way, and although it is using the scipy ckdtree functionality under the hood, the algorithm itself is written in Python. I seem to remember that a quick win was to use https://pypi.python.org/pypi/kdtree which from memory gave quite a reasonable speedup with little work, cartopy.img_transform would be the place where changes would be needed.
> 
> Cartopy's re-projection functionality is probably also paying the cost of being very general - you can provide an image in any projection, and it will put it into any other projection, dealing with discontinuities and tears without a problem. It would be really cool to hook into pyresample's functionality though (and GDAL's for that matter) to give users the opportunity to speed up the reprojection in certain cases.

Could you please provide more information how and where to start?

The kdtree library which was referenced in the answer above seems to be implemented in Python and should be slower than the scipy.ckdtree implementation (in C).

**Comment by sgdecker** [comment]
Here is some real-world code that is painfully slow, even using a very degraded resolution:
```
from datetime import datetime
import cartopy.feature as cfeature
import cartopy.crs as ccrs
import matplotlib.pyplot as plt
from metpy.plots import ctables
import numpy as np
from siphon.catalog import TDSCatalog
import xarray as xr

channel = 2
catalog_url = 'http://thredds-test.unidata.ucar.edu/thredds/catalog/'        
           'satellite/goes16/GRB16/ABI/CONUS/Channel{:02d}/current/'.format(channel)
current_goes16 = TDSCatalog(catalog_url+'catalog.xml')

latest_file = current_goes16.datasets[-1]
latest_file_url = latest_file.access_urls['OPENDAP']

ds = xr.open_dataset(latest_file_url)
rad = ds.metpy.parse_cf('Rad')
vtime = datetime.strptime(ds.time_coverage_end, '%Y-%m-%dT%H:%M:%S.%fZ')
dataproj = rad.metpy.cartopy_crs

xs = slice(3000,10000)
ys = slice(0,5000)

x = rad.x[xs].values
y = rad.y[ys].values

values = ds.Rad[ys,xs].values

ijskip = 20
x = x[::ijskip]
y = y[::ijskip]
values = values[::ijskip,::ijskip]

vis = np.sqrt(values*ds.kappa0.values)
print(vis.shape)

mapproj = ccrs.LambertConformal(central_longitude=-90.0, central_latitude=39.0,
                                standard_parallels=(30,45))


fig = plt.figure(1, figsize=(12,9), dpi=100)
ax = plt.subplot(111, projection=mapproj)

ax.add_feature(cfeature.COASTLINE.with_scale('50m'), edgecolor='sienna',
               linewidths=0.75)
ax.add_feature(cfeature.STATES.with_scale('50m'), edgecolor='sienna',
               linewidths=0.75)

img = ax.imshow(vis, extent=(x.min(), x.max(), y.min(), y.max()),
                origin='upper', cmap=plt.cm.Greys_r, transform=dataproj)

ax.set_extent((-105,-60,20,55))

plt.title('GOES-16 Visible Imagery ({:0.1f} um)'.format(ds.band_wavelength.data[0]), loc='left')
plt.title('{0:%Y-%m-%d %H:%M:%S}Z'.format(vtime), loc='right')

plt.tight_layout()
plt.show()
```

Running this through the profiler shows SciPy isn't fast enough for this case:
```
         3443291 function calls (3399468 primitive calls) in 263.840 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   1203/1    0.035    0.000  263.849  263.849 {built-in method builtins.exec}
        1    0.001    0.001  263.849  263.849 SatellitePlot.py:1(<module>)
        1    0.000    0.000  231.766  231.766 geoaxes.py:1064(imshow)
        1    0.005    0.005  231.756  231.756 img_transform.py:122(warp_array)
        1    0.034    0.034  231.745  231.745 img_transform.py:220(regrid)
        1  228.093  228.093  228.093  228.093 {method 'query' of 'scipy.spatial.ckdtree.cKDTree' objects}
8565/8453    0.207    0.000   21.019    0.002 {built-in method numpy.core.multiarray.array}
1726/1650    0.001    0.000   20.966    0.013 numeric.py:433(asarray)
    17/12    0.000    0.000   20.955    1.746 indexing.py:506(__array__)
```

**Comment by pelson** [comment]
Just as an update to this: a PR in v0.17.0, #1150, hooked us into pykdtree and ended up giving us some pretty massive straight-line speedups.  I've got some work in the wings that allows us to cache the KDTree for significant re-draw speedups. On the whole the picture has changed dramatically.

I'm closing this one out as we have actually made a huge difference to performance by using pykdtree. Obviously still plenty of room for improvement, and any early/prototype PRs that contribute in that direction are really welcome.

Cheers,
